src/USER-MISC/fix_ffl.cpp

```diff
@@ -0,0 +1,527 @@
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   http://lammps.sandia.gov, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Fast-forward Langevin thermostat, see
+   M. Hijazi, D. M. Wilkins, M. Ceriotti, J. Chem. Phys. 148, 184109 (2018)
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing authors: Lionel Constantin (EPFL), David M. Wilkins (EPFL), 
+			 Michele Ceriotti (EPFL)
+------------------------------------------------------------------------- */
+
+#include <mpi.h>
+#include <math.h>
+#include <string.h>
+#include <stdlib.h>
+#include "fix_ffl.h"
+#include "math_extra.h"
+#include "atom.h"
+#include "atom_vec_ellipsoid.h"
+#include "force.h"
+#include "update.h"
+#include "modify.h"
+#include "compute.h"
+#include "domain.h"
+#include "region.h"
+#include "respa.h"
+#include "comm.h"
+#include "input.h"
+#include "variable.h"
+#include "random_mars.h"
+#include "memory.h"
+#include "error.h"
+#include "group.h"
+
+using namespace LAMMPS_NS;
+using namespace FixConst;
+
+enum {NOBIAS,BIAS};
+enum {CONSTANT,EQUAL,ATOM};
+//#define FFL_DEBUG 1
+
+#define MAXLINE 1024
+
+/* syntax for fix_ffl:
+ * fix nfix id-group ffl tau Tstart Tstop seed [flip_type]
+ *
+ *                                                                        */
+
+/* ---------------------------------------------------------------------- */
+
+
+FixFFL::FixFFL(LAMMPS *lmp, int narg, char **arg) :
+    Fix(lmp, narg, arg)
+{
+
+
+    if (narg < 7)
+        error->all(FLERR,"Illegal fix ffl command. Expecting: fix <fix-ID>"
+                   " <group-ID> ffl <tau> <Tstart> <Tstop> <seed>  ");
+
+    restart_peratom = 1;
+    time_integrate = 1;
+    scalar_flag = 1;
+
+    //gamma = 1/ time constant(tau)
+    if (force->numeric(FLERR,arg[3]) <= 0) 
+        error->all(FLERR,"Illegal fix ffl tau value, should be greater than 0");
+    gamma = 1.0/force->numeric(FLERR,arg[3]);
+    ffl_every=1;
+    ffl_step=0;
+
+    // start temperature (t ramp)
+    t_start = force->numeric(FLERR,arg[4]);
+
+    // final temperature (t ramp)
+    t_stop = force->numeric(FLERR,arg[5]);
+
+    // PRNG seed
+    int seed = force->inumeric(FLERR,arg[6]);
+
+    // Flip type used, uses rescale if no flip is given
+    if (narg == 8) strcpy(flip_type,arg[7]);
+    else strcpy(flip_type,"rescale");
```

Comment:
Please use strncpy() here or make flip_type dynamically allocated, so that flip_type does not overflow and cause a crash on incorrect input.

Reply:
... or use an enumerator as it is done in many similar cases in LAMMPS.